File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return dfs(root).first;
    }

private:
    pair<bool, int> dfs(TreeNode* node) {
        if(node == nullptr)
            return {true, 0};
        pair<bool, int> left = dfs(node->left);
        pair<bool, int> right = dfs(node->right);

        int height = max(left.second, right.second) + 1;
        bool balanced = left.first and right.first and (abs(left.second - right.second) <= 1);

        return {balanced, height};
    }
    
};

File: Data Structures & Algorithms/path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def dfs(root, target):
            if not root:
                return False

            
            temp = target - root.val
            if not (root.left or root.right):
                return temp == 0

            return dfs(root.left, temp) or dfs(root.right, temp)
        
        return dfs(root, targetSum)